clients/importapi/importapi.go

```diff
@@ -0,0 +1,246 @@
+	path := api.url + "/jobs/" + jobID
+	jsonUpload := []byte(`{"state":"` + newState + `"}`)
+
+	jsonResult, httpCode, err := api.putJson(ctx, path, 0, jsonUpload)
+	logData := log.Data{
+		"path":        path,
+		"importJobID": jobID,
+		"jsonUpload":  jsonUpload,
+		"httpCode":    httpCode,
+		"jsonResult":  jsonResult,
+	}
+	if err == nil && httpCode != http.StatusOK {
+		err = errors.New("Bad HTTP response")
+	}
+	if err != nil {
+		log.ErrorC("UpdateImportJobState", err, logData)
+		return err
+	}
+	return nil
+}
+
+func (api *Client) getJson(ctx context.Context, path string, attempts int, vars url.Values) ([]byte, int, error) {
+	return callJsonAPI(ctx, api.client, "GET", path, vars)
+}
+
+func (api *Client) putJson(ctx context.Context, path string, attempts int, payload []byte) ([]byte, int, error) {
+	return callJsonAPI(ctx, api.client, "PUT", path, payload)
+}
+
+// func callJsonAPI(ctx context.Context, client *rchttp.Client, method, path string, payload interface{}) ([]byte, int, error) {
+func callJsonAPI(ctx context.Context, // client *rchttp.Client
+	client common.RCHTTPClienter, method, path string, payload interface{}) ([]byte, int, error) {
+
+	logData := log.Data{"url": path, "method": method}
+
+	URL, err := url.Parse(path)
+	if err != nil {
+		log.ErrorC("Failed to create url for API call", err, logData)
+		return nil, 0, err
+	}
+	path = URL.String()
+	logData["url"] = path
+
+	var req *http.Request
+
+	if payload != nil && method != "GET" {
+		req, err = http.NewRequest(method, path, bytes.NewReader(payload.([]byte)))
+		req.Header.Add("Content-type", "application/json")
+		logData["payload"] = string(payload.([]byte))
+	} else {
+		req, err = http.NewRequest(method, path, nil)
+
+		if payload != nil && method == "GET" {
+			req.URL.RawQuery = payload.(url.Values).Encode()
+			logData["payload"] = payload.(url.Values)
+		}
+	}
+	// check above req had no errors
+	if err != nil {
+		log.ErrorC("Failed to create request for API", err, logData)
+		return nil, 0, err
+	}
+
+	resp, err := client.Do(ctx, req)
+	if err != nil {
+		log.ErrorC("Failed to action API", err, logData)
+		return nil, 0, err
+	}
+
+	logData["httpCode"] = resp.StatusCode
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= 300 {
+		log.Debug("unexpected status code from API", logData)
+	}
+
+	jsonBody, err := getBody(resp)
+	if err != nil {
+		log.ErrorC("Failed to read body from API", err, logData)
+		return nil, resp.StatusCode, err
+	}
+	return jsonBody, resp.StatusCode, nil
+}
+
+// NewAPIResponse creates an error response, optionally adding body to e when status is 404
+func NewAPIResponse(resp *http.Response, uri string) (e *ErrInvalidAPIResponse) {
+	e = &ErrInvalidAPIResponse{
+		actualCode: resp.StatusCode,
+		uri:        uri,
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		body, err := getBody(resp)
+		if err != nil {
+			e.body = "Client failed to read Client body"
```

Comment:
failed to read `client` or `response` body?

Reply:
👍